src/main/java/org/opentripplanner/updater/vehicle_positions/VehiclePositionPatternMatcher.java

```diff
@@ -0,0 +1,195 @@
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import org.opentripplanner.common.model.T2;
+import org.opentripplanner.model.FeedScopedId;
+import org.opentripplanner.model.StopLocation;
+import org.opentripplanner.model.Trip;
+import org.opentripplanner.model.TripPattern;
+import org.opentripplanner.model.vehicle_position.RealtimeVehiclePosition;
+import org.opentripplanner.model.vehicle_position.RealtimeVehiclePosition.StopStatus;
+import org.opentripplanner.routing.services.RealtimeVehiclePositionService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Responsible for converting vehicle positions in memory to exportable ones, and associating each
+ * position with a pattern.
+ */
+public class VehiclePositionPatternMatcher {
+
+    private static final Logger LOG =
+            LoggerFactory.getLogger(VehiclePositionPatternMatcher.class);
+
+    private final String feedId;
+    private final RealtimeVehiclePositionService service;
+
+    private final Supplier<Map<FeedScopedId, Trip>> getTripForId;
+    private final Supplier<Map<Trip, TripPattern>> getPatternForTrip;
+
+    private Set<TripPattern> patternsInPreviousUpdate = Set.of();
+
+    public VehiclePositionPatternMatcher(
+            String feedId,
+            Supplier<Map<FeedScopedId, Trip>> getTripForId,
+            Supplier<Map<Trip, TripPattern>> getPatternForTrip,
+            RealtimeVehiclePositionService service
+    ) {
+        this.feedId = feedId;
+        this.getTripForId = getTripForId;
+        this.getPatternForTrip = getPatternForTrip;
+        this.service = service;
+    }
+
+    /**
+     * Attempts to match each vehicle position to a pattern, then adds each to a pattern
+     *
+     * @param vehiclePositions List of vehicle positions to match to patterns
+     */
+    public void applyVehiclePositionUpdates(List<VehiclePosition> vehiclePositions) {
+
+        var positions = vehiclePositions.stream()
+                .map(vehiclePosition -> toRealtimeVehiclePosition(feedId, vehiclePosition))
+                .filter(Objects::nonNull)
+                .collect(Collectors.groupingBy(t -> t.first))
+                .entrySet()
+                .stream()
+                .collect(Collectors.toMap(
+                        Entry::getKey,
+                        e -> e.getValue().stream().map(t -> t.second).collect(Collectors.toList())
+                ));
+
+        positions.forEach(service::setVehiclePositions);
+        Set<TripPattern> patternsInCurrentUpdate = positions.keySet();
+
+        // if there was a position in the previous update but not in the current one, we assume
+        // that the pattern has no more vehicle positions.
+        var toDelete = Sets.difference(patternsInPreviousUpdate, patternsInCurrentUpdate);
+        toDelete.forEach(service::clearVehiclePositions);
+        patternsInPreviousUpdate = patternsInCurrentUpdate;
```

Comment:
Should we just have a map keyed by the feed, and replace that whole map at once?

Reply:
We could but then lose the ability to have two updaters adding vehicles for the same feed. Not sure if that is a requirement but I'm a little hesitant to introduce a limitation like this.